File: daily-topic/sortColors.go
package daily_topic

func sortColors(nums []int) {
	p, n, e := 0, 0, len(nums)-1
	swap := func(n1, n2 int) {
		tmp := nums[n1]
		nums[n1] = nums[n2]
		nums[n2] = tmp
	}
	i := 0
	for i <= e {
		if nums[i] == n {
			swap(i, p)
			p++
		}
		if i == e && n != 2 {
			n++
			i = p - 1
		}
		i++
	}
}

File: host100/SortColors.java
/**
 * @author : zexho
 * @created : 2022-05-28
**/
public class SortColors{
    public void sortColors(int[] nums) {
        int left = 0, idx = 0, right = nums.length-1;
        while(idx <= right){
            if (nums[idx] == 0){
                swap(nums,idx++,left++);
            } else if (nums[idx] == 2){
                swap(nums,idx,right--);
            }else {
                idx++;
            }
        }
    }

    public void swap(int[] nums,int p1,int p2){
        int n = nums[p1];
        nums[p1] = nums[p2];
        nums[p2] = n;
    }
}
